recipes/full_finetune_single_device.py

```diff
@@ -221,18 +229,55 @@ def _setup_model(
         return model
 
     def _setup_optimizer(
-        self, cfg_optimizer: DictConfig, opt_state_dict: Optional[Dict[str, Any]] = None
-    ) -> Optimizer:
+        self,
+        cfg_optimizer: DictConfig,
+        optimizer_in_bwd: bool = False,
+        opt_state_dict: Optional[Dict[str, Any]] = None,
+    ) -> Union[Optimizer, None]:
         """
         Set up the optimizer. This method also handles loading the optimizer state_dict, if specified.
         """
-        optimizer = config.instantiate(cfg_optimizer, self._model.parameters())
-
-        if opt_state_dict:
-            optimizer.load_state_dict(opt_state_dict)
+        if optimizer_in_bwd:
+            self._optimizer_in_bwd = True
+            # Maintain a dict of optims for every parameter.
+            # TODO (rohan-varma): check foreach arg
+            optim_dict = {
+                p: config.instantiate(cfg_optimizer, [p])
+                for p in self._model.parameters()
+            }
+            self._optim_ckpt_wrapper = utils.OptimizerInBackwardWrapper(
+                {n: optim_dict[p] for n, p in self._model.named_parameters()}
+            )
 
-        log.info("Optimizer is initialized.")
-        return optimizer
+            def optim_step(param) -> None:
+                optim_dict[param].step()
+                optim_dict[param].zero_grad()
+
+            for p in self._model.parameters():
+                p.register_post_accumulate_grad_hook(optim_step)
```

Comment:
I would say put the optimizer in backward logic in a separate utility, it uses some non-intuitive logic that may confuse users

Reply:
I agree and see the reasoning in general, but IMO a downside of something like a `utils.setup_optim_in_backward(optim_config, model)` is that there is a side-effect of hooks getting registered on the model. I want things that modify state to be as explicit as possible, so I could do something like `register_optimizer_in_backward_hooks` and `make_optim_checkpoint_wrapper` - more utilities / components than a monolithic thing that configures the entire optimizer in backward. @kartikayk , @ebsmothers what do you think?